packages/mui-joy/src/Modal/useModal.ts

```diff
@@ -0,0 +1,214 @@
+'use client';
+import * as React from 'react';
+import {
+  unstable_ownerDocument as ownerDocument,
+  unstable_useForkRef as useForkRef,
+  unstable_useEventCallback as useEventCallback,
+} from '@mui/utils';
+import { ModalManager, ModalOwnProps } from '@mui/base/Modal';
+import { EventHandlers, extractEventHandlers } from '@mui/base/utils';
+
+export interface UseModalRootSlotOwnProps {
+  role: React.AriaRole;
+  onKeyDown: React.KeyboardEventHandler;
+  ref: React.RefCallback<Element> | null;
+}
+
+export interface UseModalBackdropSlotOwnProps {
+  'aria-hidden': React.AriaAttributes['aria-hidden'];
+  onClick: React.MouseEventHandler;
+  open?: boolean;
+}
+
+export type UseModalBackdropSlotProps<TOther = {}> = TOther & UseModalBackdropSlotOwnProps;
+
+export type UseModalRootSlotProps<TOther = {}> = TOther & UseModalRootSlotOwnProps;
+
+type UseModalParameters = Pick<
+  ModalOwnProps,
+  'container' | 'disableEscapeKeyDown' | 'disableScrollLock' | 'open'
+> & {
+  'aria-hidden'?: React.AriaAttributes['aria-hidden'];
+  onClose?: {
+    bivarianceHack(event: {}, reason: 'backdropClick' | 'escapeKeyDown' | 'closeClick'): void;
+  }['bivarianceHack'];
+  onKeyDown?: React.KeyboardEventHandler;
+  ref: React.Ref<HTMLElement>;
+};
+
+function ariaHidden(element: Element, show: boolean): void {
+  if (show) {
+    element.setAttribute('aria-hidden', 'true');
+  } else {
+    element.removeAttribute('aria-hidden');
+  }
+}
+
+function getContainer(container: UseModalParameters['container']) {
+  return (typeof container === 'function' ? container() : container) as HTMLElement;
+}
+
+// A modal manager used to track and manage the state of open Modals.
+// Modals don't open on the server so this won't conflict with concurrent requests.
+const manager = new ModalManager();
+
+const useModal = (parameters: UseModalParameters) => {
```

Comment:
To move in Base UI?

Reply:
In the future yes, we will. The idea was to add it in Joy UI first and validate it, and later move it to Base UI.

Reply:
Ok fair enough, I was thinking of exposing it as unstable from Base UI, but you own the Modal, so I think we are good, there is little risk that this gets duplicated with Material UI.

Things I find confusing:

1. A Joy UI Modal is not equivalent to a Base UI Modal, it's equivalent to a Material UI Modal, but Dialog. As far as know, this is not the right terminology from an a11y standpoint. See the 3 different notions:
  - Dialog: https://www.w3.org/WAI/ARIA/apg/patterns/alertdialog/
  - Modal Dialog: https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/
  - Modal: https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/

Overall, I feel that Joy UI Modal contradicts what we say in https://mui.com/base-ui/react-modal/

> The term "modal" is sometimes used interchangeably with "dialog," but this is incorrect. A dialog may be modal or nonmodal (modeless).

A Joy UI Modal doesn't have to block interactions with the rest of the page.

2. The need for a hook. In Material UI component composition was enough. Is the objective to improve performance?

Reply:
Actually, I am moving it to Base UI and marking it as `unstable`.

Reply:
The initial need for the hook was adding a `Drawer` component in Joy UI. The `Drawer` is basically a modal with different transitions. I started with implementing it using component composition, however, we discussed that probably extracting the logic in a hook would be better, so that we avoid some of the problems we have in Material UI with component composition (see https://github.com/mui/material-ui/pull/38169) - this is why I initially opened the pull request. However, it seems better to have this hook on Base UI level, so that long-term Material UI will benefit it too, both in terms of performance and avoiding some of the problems that naturally comes from component composition. We can base all Modal-like component to use the hook instead of using component composition.

I am moving the hook to Base UI as unstable.

> A Joy UI Modal doesn't have to block interactions with the rest of the page. How about we rename Joy UI Modal to Dialog?

Actually, this is exactly the same behavior as Base UI - the Modal component was literally copy pasted. The Modals both in Base UI and Joy UI do not interrupt interactions - you can make the modal interrupt interactions by using the `onClose`'s reason 'backdropClick' and not react to it. This applies even in v4 Modal - https://v4.mui.com/components/modal/.

Reply:
I remember an issue we got recently about someone complaining that if the backdrop is not used in some of the modal components, it doesn't close on clicking outside - https://github.com/mui/material-ui/issues/37760. Agree, we have a wrong description & likely people have wrong expectations based on how the component work. I will dug deeper into this.

Reply:
> A Joy UI Modal is not equivalent to a Base UI Modal, it's equivalent to a Material UI Dialog

@oliviertassinari Can you elaborate? From my view, Joy UI Modal is the same as Base UI Modal. You can pass any children to create blocking interaction to let users focus on the content of the Modal.

Now Joy UI also provides a `ModalDialog` which follows https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/

> A Joy UI Modal doesn't have to block interactions with the rest of the page. How about we rename Joy UI Modal to Dialog?

Joy UI Modal **should** block interactions. What we need to do is create a new component called `Dialog` which will not block interaction by default and let developers integrate with their triggers.

> The Joy UI's Modal page can likely be split into multiple pages, one talking about the modal and one talking about Dialogs (and alerts). It will be easier to distinguish what is modal vs dialog

Agree, we can consider splitting the page when we have a `Dialog` component for Joy UI.

Reply:
@siriwatknp we discussed this offline with Olivier this comment https://github.com/mui/material-ui/pull/38187#discussion_r1283034101 summarized the conclusions. Everything is fine with the Joy UI's modal, we just likely need to split the page so that it's easier to consume :) 

The other next thing would be supporting modal and non-modal version for all components that have some kind of drop downs, selects, autocomplete, menu etc.

Reply:
> However, we should support both of these scenarios in both Material UI and Joy UI.

Oh, yes, this would be so great 👍 

> We can consider splitting the page

This would make for a higher level of clarity 💯

> What we need to do is create a new component called Dialog which will not block interaction by default and let developers integrate with their triggers.

I'm not sure, what would a "non modal dialog" look like? Something like this https://opensource.ebay.com/mindpatterns/messaging/toast-dialog/index.html? 

I think that https://mui.com/joy-ui/react-modal/#alert-dialog is wrong, it shouldn't close when clicking on the backdrop. https://www.notion.so/mui-org/Alert-Dialog-should-close-on-backdrop-click-443f21d45852469bb3c71d94be6a57e6